### PR TITLE
[WiP] Support and enable supporting more webhooks

### PR DESCRIPTION
**Pull request by fbelzunc**

### CODE REVIEW

**Comment by talcloudshare** [comment]
Any estimates when this should be merge?

If I understand correctly, this adds the ability to have pull request triggers (as well as other triggers), which is something I'm looking for

**Comment by douglasmiller** [comment]
+1 to @talcloudshare's question. I've also made PR #32, which adds additional functionality to this PR.

**Comment by fbelzunc** [comment]
@talcloudshare @douglasmiller For the moment, the problem with this PR is that it totally change the behaviour of the plugin and for the moment it doesn't keep the job-dsl compatibility it had before and also it doesn't migrate the data from the old structure to the new one. So people with 2000 jobs with bitbucket plugin installed will lose the bitbucket configuration after installing this.

I could try to fix this, but unfortunately for the moment I didn't find spare time to look at this.

**Comment by fbelzunc** [comment]
FYI I already started working on this - just added the `readResolve`stuff.

**Comment by douglasmiller** [comment]
Great news! I've been watching the changes that you're making and will try to find some time to update my own PR with as many of the same types of changes.

I was at a total loss with the job-dsl part, but I can mimic the changes that you made for the other actions there.

**Comment by Kanti** [comment]
Please release the fix.
@see https://issues.jenkins-ci.org/browse/JENKINS-28877?focusedCommentId=265025&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-265025

**Comment by fbelzunc** [comment]
@Kanti More things need to be done - it is not ready for a release.

I am currently testing it and I think I found already a regression.

**Comment by Kanti** [comment]
@fbelzunc Thank you for the fast replay 
Can you give me a link or a brief description how to install a SNAPSHOT?
I'm not a Java guy.

**Comment by fbelzunc** [comment]
@Kanti here you have it

snapshot

Right now, there is a regression which produces each push on bitbucket to trigger all the jobs independently of the branch you want to trigger. I am fixing it, but unfortunately I go on holidays today and will not have time in 2 weeks. I hope to get this released at the end of September. That is my goal.

**Comment by rnelson0** [comment]
Any update on this? I'm running into the same [JENKINS-28877](https://issues.jenkins-ci.org/browse/JENKINS-28877) issue, would be great to have a fix for this.

**Comment by wilreichert** [comment]
Another vote for releasing this. The code in master is relatively useless now for anyone running server as the user agent check in the payload processor is [incorrect]( https://github.com/jenkinsci/bitbucket-plugin/blob/master/src/main/java/com/cloudbees/jenkins/plugins/BitbucketPayloadProcessor.java#L23)

bitbucket-webhooks
```
POST /bitbucket-hook/ HTTP/1.1 
Content-Type: application/json 
User-Agent: Bitbucket 4.14.3 
X-Atlassian-Token: no-check 
Accept: application/json 
Content-Length: 9754 
Host: almond.mac.local:8080 
Connection: Keep-Alive 
Accept-Encoding: gzip,deflate 
 ```

Bitbucket Web Post Hooks Plugin
```
POST /bitbucket-hook/ HTTP/1.1 
X-Bitbucket-Type: server 
X-Event-Key: repo:push 
Content-Type: application/json; charset=UTF-8 
Via: 1.1 localhost (Apache-HttpClient/4.4.1 (cache)) 
Content-Length: 494 
Host: almond.mac.local:8080 
Connection: Keep-Alive 
User-Agent: Atlassian HttpClient 0.23.0 / Bitbucket-4.14.3 (4014003) / Default 
```

**Comment by gravityuser** [comment]
+1+1+1+1+1+1+1+1+1+1

**Comment by gravityuser** [comment]
This would be so useful and so many open PRs are to enable this kind of functionality.  We can help support/build this out as this project is stale.

**Comment by rnelson0** [comment]
@fbelzunc Any update on the ETA at least?

**Comment by fbelzunc** [comment]
This PR still has a lot of problems - it cannot be merged. However, I don't plan to continue working on this for the moment due lack of time with the ad-plugin... if someone address all the problems it can be merged.

Notice that nothing stop you of creating a SNAPSHOT and installing it.

**Comment by rnelson0** [comment]
@fbelzunc Well, it kind of does. I'm a tenant on a centralized Jenkins service and they'll only install plugins that are released through the normal channels, they won't install a tgz from [here](https://jenkins.ci.cloudbees.com/job/plugins/job/bitbucket-plugin/141/org.jenkins-ci.plugins$bitbucket/).

But, to get this merged, is there a list of items that need fixed that we could contribute toward? I don't know anything about Java but it looks like there's definitely others interested who may be able to contribute.

**Comment by douglasmiller** [comment]
@fbelzunc I'd be willing to spend time addressing any issues that I'm capable of fixing. Do you have a list of the issues with this change anywhere? I have my own PR that's targeting this develop branch which supports more of the Bitbucket webhooks.

**Comment by fbelzunc** [comment]
I plan to work on this very soon.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketHookReceiver.java [hunk 1]
```diff
@@ -51,7 +78,16 @@ public void doIndex(StaplerRequest req) throws IOException {
...
 
-            payloadProcessor.processPayload(payload, req);
+            if ("Bitbucket-Webhooks/2.0".equals(req.getHeader("user-agent"))) {
+                BitbucketEvent bitbucketEvent = new BitbucketEvent(req.getHeader("x-event-key"));
+                BitbucketPayloadProcessor bitbucketPayloadProcessor = payloadProcessorFactory.create(bitbucketEvent);
+                bitbucketPayloadProcessor.processPayload(payload);
+            } else {
+                LOGGER.log(Level.INFO, "Processing old POST service payload");
+                BitbucketPayloadProcessor bitbucketPayloadProcessor =
+                        payloadProcessorFactory.createOldProcessor(new BitbucketEvent("repo:push"));
```
Comment (krotte1) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketHookReceiver.java]: This implementation does not support on-prem implementations of Bitbucket server where the payloads (see (https://confluence.atlassian.com/bitbucketserver0512/event-payload-955159824.html) are different. Check out pull-request 58 and try to incorporate some of those changes.

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+/**
+ * Bitbucket events
+ * @since August 1, 2016
+ * @version 2.0
+ */
+public class BitbucketEvent {
+    public interface EVENT {
+        String REPOSITORY = "repo";
+        String PULL_REQUEST = "pullrequest";
```
Comment (krotte1) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: The string used for Bitbucket server (on-prem) is different:
PULL_REQUEST = "pr"

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    public interface REPOSITORY_ACTIONS {
+        String PUSH = "push";
+    }
+
+    public interface PULL_REQUEST_ACTIONS {
+        String CREATED = "created";
```
Comment (krotte1) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Bitbucket Server values are different:
CREATED = "opened"

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        String APPROVED = "approved";
```
Comment (krotte1) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Bitbucket server... when a user approves a PR, the event is:
APPROVED = "reviewer:approved"

File: src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        }
+    }
+
+    /**
+     * Returns {@code true} if the action is defined
+     *
+     * @return {@code true} if the action is defined
+     */
+    private boolean checkAction(String action) {
+        if(EVENT.REPOSITORY.equals(name)) {
```
Comment (krotte1) [inline src/main/java/com/cloudbees/jenkins/plugins/BitbucketEvent.java]: Needs to be expanded to support the Bitbucket server variables mentioned earlier.

**Comment by tzachs** [comment]
@fbelzunc with your permission, I think I'm going to close this one without merging. The reason is the trust level and there a lot of changes here.
WDYT?

**Comment by tzachs** [comment]
Closing this one as it still WIP and the plug-in has changed a lot due to upgrade to base Jenkins version 2.204.1

**Pull request closed without merging**